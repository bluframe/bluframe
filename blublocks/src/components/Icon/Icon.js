// @flow

import type { ComponentProps } from "."
import React from "react"
import { Icon as StyledIcon } from "./styled"

const Icon = ({ color, SelectedIcon, size }: ComponentProps): React$Node => (
  <StyledIcon>
    <SelectedIcon color={color} size={size} />
  </StyledIcon>
)

export default Icon
