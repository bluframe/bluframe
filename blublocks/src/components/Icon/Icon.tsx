import { ComponentProps } from "."

const Icon = ({
  ariaHidden,
  color,
  SelectedIcon,
  size = "16px",
  title
}: ComponentProps) => (
  <SelectedIcon
    aria-hidden={ariaHidden}
    color={color}
    size={size}
    title={title}
  />
)

export default Icon
