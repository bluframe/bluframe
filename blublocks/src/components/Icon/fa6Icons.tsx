import {
  FaFacebook,
  FaInstagram,
  FaLinkedin,
  FaPencil,
  FaPhone,
  FaQuoteLeft,
  FaQuoteRight,
  FaReddit,
  FaRegFileLines,
  FaXTwitter,
  FaYoutube
} from "react-icons/fa6"

const fa6Icons = Object.freeze({
  FaFacebook,
  FaInstagram,
  FaLinkedin,
  FaPencil,
  FaPhone,
  FaQuoteLeft,
  FaQuoteRight,
  FaReddit,
  FaRegFileLines,
  FaXTwitter,
  FaYoutube
})

export default fa6Icons
