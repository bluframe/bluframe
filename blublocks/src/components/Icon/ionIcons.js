// @flow

import { IoAccessibilityOutline } from "@react-icons/all-files/io5/IoAccessibilityOutline"
import { IoAdd } from "@react-icons/all-files/io5/IoAdd"
import { IoAddCircle } from "@react-icons/all-files/io5/IoAddCircle"
import { IoAlertCircleOutline } from "@react-icons/all-files/io5/IoAlertCircleOutline"
import { IoArrowBack } from "@react-icons/all-files/io5/IoArrowBack"
import { IoArrowForward } from "@react-icons/all-files/io5/IoArrowForward"
import { IoBagOutline } from "@react-icons/all-files/io5/IoBagOutline"
import { IoBookmarkOutline } from "@react-icons/all-files/io5/IoBookmarkOutline"
import { IoBulbOutline } from "@react-icons/all-files/io5/IoBulbOutline"
import { IoCaretUp } from "@react-icons/all-files/io5/IoCaretUp"
import { IoChatbubbleOutline } from "@react-icons/all-files/io5/IoChatbubbleOutline"
import { IoCheckmark } from "@react-icons/all-files/io5/IoCheckmark"
import { IoClose } from "@react-icons/all-files/io5/IoClose"
import { IoCloudOutline } from "@react-icons/all-files/io5/IoCloudOutline"
import { IoEaselOutline } from "@react-icons/all-files/io5/IoEaselOutline"
import { IoHeartOutline } from "@react-icons/all-files/io5/IoHeartOutline"
import { IoHelpOutline } from "@react-icons/all-files/io5/IoHelpOutline"
import { IoHomeOutline } from "@react-icons/all-files/io5/IoHomeOutline"
import { IoLaptopOutline } from "@react-icons/all-files/io5/IoLaptopOutline"
import { IoLogoFacebook } from "@react-icons/all-files/io5/IoLogoFacebook"
import { IoLogoTwitter } from "@react-icons/all-files/io5/IoLogoTwitter"
import { IoMapOutline } from "@react-icons/all-files/io5/IoMapOutline"
import { IoPaperPlaneOutline } from "@react-icons/all-files/io5/IoPaperPlaneOutline"
import { IoPencilOutline } from "@react-icons/all-files/io5/IoPencilOutline"
import { IoPeopleOutline } from "@react-icons/all-files/io5/IoPeopleOutline"
import { IoPhonePortraitOutline } from "@react-icons/all-files/io5/IoPhonePortraitOutline"
import { IoSearch } from "@react-icons/all-files/io5/IoSearch"
import { IoSettingsOutline } from "@react-icons/all-files/io5/IoSettingsOutline"
import { IoShareSocialOutline } from "@react-icons/all-files/io5/IoShareSocialOutline"
import { IoSpeedometerOutline } from "@react-icons/all-files/io5/IoSpeedometerOutline"
import { IoTrendingDownOutline } from "@react-icons/all-files/io5/IoTrendingDownOutline"
import { IoTrendingUpOutline } from "@react-icons/all-files/io5/IoTrendingUpOutline"
import { IoWarningOutline } from "@react-icons/all-files/io5/IoWarningOutline"

const ionIcons = Object.freeze({
  IoAccessibilityOutline,
  IoAdd,
  IoAddCircle,
  IoAlertCircleOutline,
  IoArrowBack,
  IoArrowForward,
  IoBagOutline,
  IoBookmarkOutline,
  IoBulbOutline,
  IoCaretUp,
  IoChatbubbleOutline,
  IoCheckmark,
  IoClose,
  IoCloudOutline,
  IoEaselOutline,
  IoHeartOutline,
  IoHelpOutline,
  IoHomeOutline,
  IoLaptopOutline,
  IoLogoFacebook,
  IoLogoTwitter,
  IoMapOutline,
  IoPaperPlaneOutline,
  IoPencilOutline,
  IoPeopleOutline,
  IoPhonePortraitOutline,
  IoSearch,
  IoSettingsOutline,
  IoShareSocialOutline,
  IoSpeedometerOutline,
  IoTrendingDownOutline,
  IoTrendingUpOutline,
  IoWarningOutline
})

export default ionIcons
