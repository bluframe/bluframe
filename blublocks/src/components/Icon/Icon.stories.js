// @flow

import Icon, { type Props } from "components/Icon"
import React from "react"
import type { Story } from "@storybook/react"

const stories = {
  argTypes: {
    name: {
      control: {
        options: [
          "IoAccessibilityOutline",
          "IoAdd",
          "IoArrowBack",
          "IoArrowForward",
          "IoBagOutline",
          "IoBookmarkOutline",
          "IoBulbOutline",
          "IoCaretUp",
          "IoChatbubbleOutline",
          "IoCheckmark",
          "IoClose",
          "IoCloudOutline",
          "IoEaselOutline",
          "IoHeartOutline",
          "IoHelpOutline",
          "IoHomeOutline",
          "IoLaptopOutline",
          "IoLogoFacebook",
          "IoLogoTwitter",
          "IoMapOutline",
          "IoPaperPlaneOutline",
          "IoPencilOutline",
          "IoPeopleOutline",
          "IoPhonePortraitOutline",
          "IoSearch",
          "IoSettingsOutline",
          "IoShareSocialOutline",
          "IoSpeedometerOutline",
          "IoTrendingDownOutline",
          "IoTrendingUpOutline",
          "IoWarningOutline"
        ],
        type: "select"
      }
    }
  },
  component: Icon,
  title: "Components/Icon"
}

const Template = (args: Props) => <Icon {...args} />

export const Ion: Story<Props> = Template.bind({})
Ion.args = {
  group: "io5",
  name: "IoSpeedometerOutline",
  size: "50px"
}

export default stories
