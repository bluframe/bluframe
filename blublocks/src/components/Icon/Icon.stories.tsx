import { Meta, StoryObj } from "@storybook/react"
import Icon from "components/Icon"
import fa6Icons from "./fa6Icons"
import ionIcons from "./ionIcons"

const meta: Meta<typeof Icon> = {
  argTypes: {
    color: { control: "color" }
  },
  component: Icon,
  title: "Components/Icon"
}

type Story = StoryObj<typeof Icon>

export const Ion: Story = {
  argTypes: {
    name: {
      control: "select",
      options: Object.keys(ionIcons)
    }
  },
  args: {
    group: "io5",
    name: "IoSpeedometerOutline",
    size: "50px"
  }
}

export const Fa: Story = {
  argTypes: {
    name: {
      control: "select",
      options: Object.keys(fa6Icons)
    }
  },
  args: {
    group: "fa6",
    name: "FaQuoteLeft",
    size: "50px"
  }
}

export default meta
