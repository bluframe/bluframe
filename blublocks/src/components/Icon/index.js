// @flow

import type { IconProps, SelectedIconProps } from "@bluframe/blublocks"
import Icon from "./Icon"
import ionIcons from "./ionIcons"
import { prepareComponent } from "@bluframe/grapple"

export type Props = {|
  ...IconProps
|}

export type ComponentProps = {|
  ...SelectedIconProps,
  +SelectedIcon: React$ComponentType<SelectedIconProps>
|}

const usePrepareComponent = ({
  color,
  group = "io5",
  name,
  size
}: Props): ComponentProps => {
  if (group === "io5" && ionIcons[name]) {
    return {
      SelectedIcon: ionIcons[name],
      color,
      size
    }
  }

  return {
    SelectedIcon: () => null,
    color,
    size
  }
}

export default (prepareComponent<Props, ComponentProps>(usePrepareComponent)(
  Icon
): React$ComponentType<Props>)
