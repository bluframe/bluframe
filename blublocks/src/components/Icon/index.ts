import Icon from "./Icon"
import { IconType } from "react-icons"
import fa6Icons from "./fa6Icons"
import ionIcons from "./ionIcons"
import { prepareComponent } from "@bluframe/grapple"

export type IconGroupName = "fa6" | "io5"

export interface Props {
  ariaHidden?: boolean
  color?: string
  group?: IconGroupName
  name: keyof typeof fa6Icons | keyof typeof ionIcons
  size?: string
  title?: string
}

export interface ComponentProps extends Omit<Props, "group" | "name"> {
  SelectedIcon: IconType | (() => null)
}

const isIonIcon = (name: Props["name"]): name is keyof typeof ionIcons => {
  return name in ionIcons
}

const isFaIcon = (name: Props["name"]): name is keyof typeof fa6Icons => {
  return name in fa6Icons
}

const usePrepareComponent = ({
  group = "io5",
  name,
  ...props
}: Props): ComponentProps => {
  if (group === "io5" && isIonIcon(name)) {
    return {
      ...props,
      SelectedIcon: ionIcons[name]
    }
  }

  if (group === "fa6" && isFaIcon(name)) {
    return {
      ...props,
      SelectedIcon: fa6Icons[name]
    }
  }

  return {
    ...props,
    SelectedIcon: () => null
  }
}

export default prepareComponent<Props, ComponentProps>(usePrepareComponent)(
  Icon
)
