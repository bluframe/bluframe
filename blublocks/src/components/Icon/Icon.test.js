// @flow

import Icon from "./Icon"
import React from "react"
import { render } from "tests"

describe("Icon", () => {
  it("renders default size", () => {
    const props = {
      SelectedIcon: () => "SelectedIcon"
    }

    const { container } = render(<Icon {...props} />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders size", () => {
    const props = {
      SelectedIcon: () => "SelectedIcon",
      size: "32px"
    }

    const { container } = render(<Icon {...props} />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders color", () => {
    const props = {
      SelectedIcon: () => "SelectedIcon",
      color: "blue"
    }

    const { container } = render(<Icon {...props} />)

    expect(container.firstChild).toMatchSnapshot()
  })
})
