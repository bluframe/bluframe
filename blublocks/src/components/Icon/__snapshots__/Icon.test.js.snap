// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon renders color 1`] = `
<span
  class="styled__Icon-sc-ztr6p6-0 kHWcXr"
>
  SelectedIcon
</span>
`;

exports[`Icon renders default size 1`] = `
<span
  class="styled__Icon-sc-ztr6p6-0 kHWcXr"
>
  SelectedIcon
</span>
`;

exports[`Icon renders size 1`] = `
<span
  class="styled__Icon-sc-ztr6p6-0 kHWcXr"
>
  SelectedIcon
</span>
`;
